frame/collective/src/lib.rs

```diff
@@ -185,69 +311,185 @@ decl_module! {
 		/// - `new_members`: The new member list. Be nice to the chain and
 		//	provide it sorted.
 		/// - `prime`: The prime member whose vote sets the default.
+		/// - `old_count`: Upper bound for the number of previous members. (Used for weight estimation.)
 		///
 		/// Requires root origin.
-		#[weight = (100_000_000, DispatchClass::Operational)]
-		fn set_members(origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>) {
+		///
+		/// # <weight>
+		/// ## Weight
+		/// - `O(MP + N)` where:
+		///   - `M` old-members-count (code- and governance-bounded)
+		///   - `N` new-members-count (code- and governance-bounded)
+		///   - `P` proposals-count (code-bounded)
+		/// - DB:
+		///   - 1 storage mutation (codec `O(M)` read, `O(N)` write) for reading and writing the members
+		///   - 1 storage read (codec `O(P)`) for reading the proposals
+		///   - `P` storage mutations (codec `O(M)`) for updating the votes for each proposal
+		///   - 1 storage write (codec `O(1)`) for deleting the old `prime` and setting the new one
+		/// - Benchmark: 0 + M * 20.47 + N * 0.109 + P * 26.29 µs (min squares analysis)
```

Comment:
The complexity is `O(MP + N)` but the formula compute `k1*M + k2*P + k3*N` shouldn't it be `k1*M*P + k2*N` actually the best could maybe do an analysis with `k1*M + k2*P + k3*N + k4*M*P` not sure how difficult it is though

Reply:
yeah I guess it should be `k1*M + k2*P + k3*N + k4*M*P`
I figured because the linear approximation is based on the worst cases, that it should only overestimate the weight, but I'm not 100% sure this is the case.

Reply:
I would be surprised if the analysis Shawn has been running could figure out the 4 factors, though. It would likely be more like `k1*M*P + k2*N`.

Reply:
@shawntabrizi Can you tell us whether the 3 param linear approximation is a worst case and whether a quadratic approximation is possible?

Reply:
benchmarks are evaluating:
* varying M with N=100 and P=100
* varying N with M=100 and P=100
* varying P with M=100 and N=100

so if we do a solution with `F= k0 + k1*M + k2*N + k3*P + k4*M*P`
So we have `k1+100*k4 = 20.47` `k2 + 100*k4 = 0.109`.

so at most k4 is 0.001 (for k2 == 0)
so for M=100 and N=100 we would have an additional 10µs which is negligible compare to `k1*100==2_000µs`

I don't know if it worth adding in the doc. But it is safe here at least.

Reply:
add the following doc comment:
```
/// Note: The complexity of `set_members` is quadratic (`O(MP + N)`), so the linear approximation
	/// of the benchmark is not always permissible. It is here, though, because the linear approximation
	/// covered the range of possible values and we estimate weight via the worst case (max paramter 
	/// values) before execution so we can be sure that we are only overestimating.
```
Is that true and understandable?

Reply:
I'm not sure this is true. For me the reason it is safe is because the coef k4 is negligible anyway with our value of M and N.

If we had M and N until 10_000 then if k4 is 0.001 it would no longer be negligible: `k1*10_000=200_000us` and `k4*10_000*10_000=100_000us`

Reply:
Damn my math are wrong the quadratic cost here is not M*N but M*P, so my proof doesn't stand anymore

Reply:
but yes your justification is correct